### ISSUE TITLE
enhancement: client/editor should use unconnected messages to check server status

### ISSUE THREAD

**Issue by lodicolo** [AscensionGameDev/Intersect-Engine#224]
**Summary**: Currently the client/editor establish full-blown connections to the server even on the main menu, where the client could be left inactive but still hogging a connection.

**Solution**:
- Client/Editor should use unconnected messages to communicate with server before login
  - Primarily server status, version verification
- Client/Editor should request full connections on login attempt
- Client should:
  - lose the connection on logout
  - request full connections on registration

**Comment by Arufonsu** [comment]
Not sure if related, but after reading "the client establish full-blown connection" i've realized that this may be related with this issue:

(1)
-Start the server.
-Asap and while it's loading, open a client.
-Server crashes.

(2)
-Start the client. - will show up the server as offline.
-Start the server. The client will update the current statys to "online" as normal
_-No issue here_

(3)
-Start the server.
-Wait until it loads properly.
-Start the client, _no issue here_, but i shouldn't be waiting for the server to fully load, players can't tell about that right?

Why i mention these 3 situations and why i think that "the client establish full-blown connection" may be the fault of the (1) issue:
in the normal basis, when online games are offline due to maintenance or updates, players will keep opening and closing their clients in order to refresh and see if their game is up, this may lead to multiple server crashes (1).

About (2): Maybe.. just maybe, the client shouldn't start up right away, unless the server its completly online.
if the server is offline, instead of the server status in the main menu screen, a notice window should pop up saying "the game server is currently offline / under maintenance / or whatever". This way, situation 3 (where we wait in order to prevent the server crash) and situation 1 could be solved, leaving us with no situation (2) anymore at this point, yet, a more stable server that's able to load up without interruptions its always better.

If anyone can help by replicating the situation (1) that i previously mentioned, then we could confirm that the issue that @lodicolo  is talking about may be related with this as well.

And yes, adding to what @lodicolo says here, the client shouldn't be hogging a connection indefinitely, an online status check should be enough upon client start up, if the server is offline, then don't allow the client to start up. Once the client is up, connections should be somewhat limited, if no registration/login/char select request is made within x minutes, the server should drop the connection and close the user's client (i've seen this in many online games by now, they do this in order to prevent unnecessary flood of packets from afkers or users who simply leave the game open in the main/login screen.

It may not seem like a big deal right now, but when it comes to scalability and a steady increase in players, then the problem will be more than real. 

Cheers <3

**Comment by jcsnider** [comment]
@Arufonsu the server crashing when clients connect too quickly after boot is a very new issue (within the past 2 weeks). I think it might have to do with Lidgren changes @lodicolo made regarding encryption/decryption performance. Once thing that you might test if you want is downgrading the lidgren project in the nuget package manager.

It's weird because even if I have all exceptions set to get caught/paused whilst debugging that never happens. I will look into the issue soon but I think it's ultimately unrelated to this issue itself.

**Comment by Arufonsu** [comment]
alright 👍🏽 didn't saw an issue opened for that one, that's why i was relating it to this.
and yeah, its weird, tried to debugg as well without much luck, guess that's something to keep and eye on eventually.

leaving that topic off from here now, since its probably not related as @jcsnider says.

**Comment by lodicolo** [comment]
@jcsnider could use good ol' fashioned printfs to diagnose it.

If the server is crashing then there's no online check that the client can do that won't crash it, the fix has to be on the server.

Depending on what the crash is caused by we may need to just be ignoring it.

**Comment by jcsnider** [comment]
For sure the fix needs to be on the server, nothing the client can do should crash it. You are correct in that it is probably something we just need to try/catch and ignore -- I just haven't been annoyed enough yet to go fix it XD

**Comment by jcsnider** [comment]
As an aside this sometimes happens without clients connecting.... not sure what it is yet.

**Comment by lodicolo** [comment]
This is done in 0.8-beta
